src/script/interpreter.cpp

```diff
@@ -1299,9 +1317,17 @@ void PrecomputedTransactionData::Init(const T& txTo, std::vector<CTxOut> spent_o
 
     // Cache is calculated only for transactions with witness
     if (txTo.HasWitness()) {
-        hashPrevouts = GetPrevoutHash(txTo);
-        hashSequence = GetSequenceHash(txTo);
-        hashOutputs = GetOutputsHash(txTo);
+        m_prevouts_hash = GetPrevoutHash(txTo);
+        hashPrevouts = SHA256Uint256(m_prevouts_hash);
```

Comment:
these are computed at the txn level, so you'd need to add some code which detects if there is no input of that type. Maybe worthwhile?

But keep in mind that taproot isn't calculating anything new, just caching the intermediate results...

Reply:
Don't `m_spent_amounts_hash` and `m_spent_scripts_hash` (below) add two extra hash operations when verifying SegWit transactions?

Reply:
ah yes those are extra, but only when haswitness. I suppose we could add a function hasv1witness?

Reply:
I've made some significant changes here; the code will now detect if a transaction needs BIP341 and/or BIP143 sighashing, and then only precompute what is needed for the existing ones.

Reply:
I think it's after this comment that the hash names began containing bip341?

I don't think it's worth changing at this point, but I don't love that naming convention because it's more clear to name them based on what they are (e.g., m_prevouts_sha256 or m_prevouts_d_sha256) than to name based on a BIP, especially since 143 and 341 are digit wise reversed it tripped me up a bit when re-reviewing.

This is also confusing because CTV & anyprevout will also use these same single hashes, so then it's not even really accurate to call them bip341 hashes -- that's actually why I noticed, because rebasing CTV on taproot needed to be adjusted for the new names.

Again, not worth changing at this point, but if you decide to rename them again later might be worth going to non bip name.